api/api-environments-v3/src/main/java/com/thoughtworks/go/apiv3/environments/EnvironmentsControllerV3.java

```diff
@@ -0,0 +1,268 @@
+        HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
+        EnvironmentConfig environmentConfig = fetchEntityFromConfig(environmentName);
+
+        environmentConfigService.deleteEnvironment(environmentConfig, currentUsername(), result);
+        return handleSimpleMessageResponse(response, result);
+    }
+
+    @Override
+    public String etagFor(EnvironmentConfig entityFromServer) {
+        if (entityFromServer instanceof MergeEnvironmentConfig || entityFromServer instanceof UnknownEnvironmentConfig) {
+            return md5Hex(valueOf(hash(entityFromServer)));
+        }
+        return entityHashingService.md5ForEntity(entityFromServer);
+    }
+
+    @Override
+    public EntityType getEntityType() {
+        return EntityType.Environment;
+    }
+
+    @Override
+    public EnvironmentConfig doFetchEntityFromConfig(String name) {
+        return environmentConfigService.getEnvironmentConfig(name);
+    }
+
+    @Override
+    public EnvironmentConfig fetchEntityFromConfig(String nameOrId) {
+        EnvironmentConfig entity = doFetchEntityFromConfig(nameOrId);
+        if (entity == null || entity.isUnknown()) {
+            throw new RecordNotFoundException(getEntityType(), nameOrId);
+        }
+        return entity;
+    }
+
+    @Override
+    public EnvironmentConfig buildEntityFromRequestBody(Request req) {
+        JsonReader jsonReader = GsonTransformer.getInstance().jsonReaderFrom(req.body());
+        return EnvironmentRepresenter.fromJSON(jsonReader);
+    }
+
+    @Override
+    public Consumer<OutputWriter> jsonWriter(EnvironmentConfig environmentConfig) {
+        return writer -> EnvironmentRepresenter.toJSON(writer, environmentConfig);
+    }
+
+    List<EnvironmentConfig> filterUnknownAndSortEnvConfigs(Set<EnvironmentConfig> envConfigSet) {
+        return envConfigSet.stream()
+                .filter(envConfig -> !envConfig.isUnknown())
+                .sorted(comparing(EnvironmentConfig::name))
+                .collect(toList());
+    }
+
+    private void haltIfEntityWithSameNameExists(EnvironmentConfig environmentConfig) {
+        HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
+        ConfigElementForEdit<EnvironmentConfig> existingEnvConfig = environmentConfigService.getMergedEnvironmentforDisplay(environmentConfig.name().toString(), result);
+        if (existingEnvConfig == null || existingEnvConfig.getConfigElement().isUnknown()) {
+            return;
+        }
+
+        environmentConfig.addError("name", format("Environment name should be unique. Environment with name '%s' already exists.", environmentConfig.name().toString()));
+        throw haltBecauseEntityAlreadyExists(jsonWriter(environmentConfig), "environment", environmentConfig.name().toString());
+    }
+
+    private String calculateEtag(Collection<EnvironmentConfig> environmentConfigs) {
+        final String environmentConfigSegment = environmentConfigs
```

Comment:
can we move this method to `entityHashingService`?

Reply:
Refactored.